### PR TITLE
Chore: Make only member of team able to run build image workflow

### PR DESCRIPTION
**Pull request by ying-jeanne**
#### Which issue(s) this PR fixes or relates to

In long term, we should check whether the workflow is initiated by a member, rather than solely confirming the author's membership status.

This modification is necessary in scenarios where an external contributor submits a PR to modify the mimir-build-image, even if the PR receives approval from a maintainer, the build image will not update automatically. In such instances, we would be required to either manually update the build image or have a member initiate the PR to ensure the update occurs.

To avoid this, we could check whether the workflow is triggered by a member, in order to able to  start the workflow manually

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pstibrany — APPROVED** [review]
Thank you for this PR.

> I believe the appropriate approach to resolve this issue should entail confirming whether the workflow is initiated by a member, rather than solely confirming the author's membership status.
> 
> This modification is imperative because in scenarios where an external contributor submits a pull request (PR) to modify the mimir-build-image, even if the PR receives approval from a maintainer, the build image will not update automatically. In such instances, we would be required to either manually update the build image or have a member initiate the PR to ensure the update occurs.
> 
> Furthermore, relying solely on the maintainer's approval for triggering the workflow run is not sufficient, as the author can introduce additional commits after obtaining approval, which can then trigger the build unexpectedly.

You lost me at "should entail confirming". I've re-read this description several times, and I'm still confused.

Thankfully I have access to our internal slack discussion. For those who don't: we want to allow running github actions for all contributors, but don't want all contributors to be able to publish new build images just by sending the PR. Hence this change.

This is a follow-up to https://github.com/grafana/mimir/pull/5335#discussion_r1241402161

**Review by aknuds1 — APPROVED** [review]

**Comment by charleskorn** [comment]
Couldn't someone who wanted to circumvent this check remove it as part of their PR? 

Or are we relying on GitHub's [workflow approval option for outside collaborators](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) being set to "require approval for all outside collaborators"? 

If the latter, then do we need this change?

**Comment by jhesketh** [comment]
> Couldn't someone who wanted to circumvent this check remove it as part of their PR?

I *think* (but would want to confirm) that we would be protected by the user not being able to access the [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).

>  For secrets stored at the environment level, you can enable required reviewers to control access to the secrets. A workflow job cannot access environment secrets until approval is granted by required approvers.

**Comment by charleskorn** [comment]
> I _think_ (but would want to confirm) that we would be protected by the user not being able to access the [secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions).
> 
> > For secrets stored at the environment level, you can enable required reviewers to control access to the secrets. A workflow job cannot access environment secrets until approval is granted by required approvers.

The Docker username and password secrets are currently set at the repository level, not the environment level. (We don't have any environments defined at present - not sure how hard it would be to set them up or if it would make other things more difficult.)

**Comment by jhesketh** [comment]
> The Docker username and password secrets are currently set at the repository level, not the environment level. (We don't have any environments defined at present - not sure how hard it would be to set them up or if it would make other things more difficult.)

Right, but I [think we're still covered](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow):

> With the exception of `GITHUB_TOKEN`, secrets are not passed to the runner when a workflow is triggered from a forked repository.

**Comment by charleskorn** [comment]
> Right, but I [think we're still covered](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#using-secrets-in-a-workflow):
> 
> > With the exception of `GITHUB_TOKEN`, secrets are not passed to the runner when a workflow is triggered from a forked repository.

In that case, it sounds like we're already covered from the perspective of not allowing anyone to build and upload arbitrary images then - do we need this PR?

However, if someone who isn't a maintainer or contributor opens a PR to modify the build image, how do we get that image built? Would we need a maintainer to manually trigger the build, and would we still have the same restrictions given the code would still be coming from a fork?

**Comment by pracucci** [comment]
> With the exception of GITHUB_TOKEN, secrets are not passed to the runner when a workflow is triggered from a forked repository.

When you open a PR on this repository, isn't the workflow triggered on the repository itself (even if the branch was from a forked repository) and so it will get the `GITHUB_TOKEN`?

File: .github/workflows/push-mimir-build-image.yml
Comment (pracucci) [inline .github/workflows/push-mimir-build-image.yml:17]: Please add a comment to explain why.

**Comment by jhesketh** [comment]
> When you open a PR on this repository, isn't the workflow triggered on the repository itself (even if the branch was from a forked repository) and so it will get the `GITHUB_TOKEN`?

The docs is saying the GITHUB_TOKEN is always available.

But I ran a quick test on a repository secret.

This [PR](https://github.com/jhesketh/forktest/pull/4) is open from a branch within the repository and is able to use the secret as seen by the *** in the output: https://github.com/jhesketh/forktest/actions/runs/6258522449/job/16992691391?pr=4#step:2:5

This [PR](https://github.com/jhesketh/forktest/pull/5) is opened from a fork, but does not have access to the secret (it is blank): https://github.com/jhesketh/forktest/actions/runs/6258527574/job/16992706304?pr=5#step:2:5

Due to the sensitivity of this, we should test the desired behavour in this repository.

**Comment by pstibrany** [comment]
> Couldn't someone who wanted to circumvent this check remove it as part of their PR?
> 
> Or are we relying on GitHub's [workflow approval option for outside collaborators](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) being set to "require approval for all outside collaborators"?
> 
> If the latter, then do we need this change?

We planned to change setting to allow running actions for all contributors (except first-time ones), after merging this PR.

Your point about just removing the check is a good one, and I don't have an answer whether it would allow publishing of docker image or not. Definitely worth checking before the merge.

**Comment by ying-jeanne** [comment]
This pull request would allow contributors to continue running the pipeline but prevent them from updating the mimir-build-image. The responsibility for building the mimir-build-image will be shifted to organization members. While we work on a more comprehensive solution for cases where contributors wish to modify the mimir-build-image and create a new version of it, I propose to let's merge this change to remove the current obstruction in the build process for contributors.

**Comment by dimitarvdimitrov** [comment]
If the only way to push a malicious build image is to open a PR, then I think we would be to quickly spot when this happens because we get notifications for all open PRs. We can delete the image. This isn't a very scalable way of dealing with the problem, but I'm also not sure how often this problem will occur. 

I think unblocking external contributors by merging this PR and allowing CI to run for repeating contributors is a reasonable temporary measure. At least until the two workflows that @ying-jeanne wants to implement are done.

**Review by dimitarvdimitrov — APPROVED** [review]

File: .github/workflows/push-mimir-build-image.yml
Comment (dimitarvdimitrov) [inline .github/workflows/push-mimir-build-image.yml:22]: this comment seems offset

**Comment by dimitarvdimitrov** [comment]
@ying-jeanne can you share your plan for the two workflows that will prevent contributors from changing the workflow template and submit a PR which publishes a build image? I would like to work on it because currently all external contributions are slowed down because of the extra approvals

**Comment by ying-jeanne** [comment]
> @ying-jeanne can you share your plan for the two workflows that will prevent contributors from changing the workflow template and submit a PR which publishes a build image? I would like to work on it because currently all external ,contributions are slowed down because of the extra approvals

Hi @dimitarvdimitrov I wanted to surmerize the 2 workflows approach yesterday, end-up going to the rabbit hole, just avoid to lose your time also on this I test the approach on my repo and compose a draft PR https://github.com/grafana/mimir/pull/6491/files. Can you give it a look? Thanks

**Comment by ying-jeanne** [comment]
Here is the draft for 2 workflows approach: https://github.com/grafana/mimir/pull/6491

**Review by aknuds1 — COMMENTED** [review]
Left a comment (on comment indentation).

**Comment by pstibrany** [comment]
We think this is an acceptable trade-off. External contributors can't publish new docker images, because actions running on PRs from forks don't have access to the secrets defined in this repository.

**Review by pstibrany — APPROVED** [review]

**Comment by pstibrany** [comment]
> > When you open a PR on this repository, isn't the workflow triggered on the repository itself (even if the branch was from a forked repository) and so it will get the `GITHUB_TOKEN`?
> 
> The docs is saying the GITHUB_TOKEN is always available.
> 
> But I ran a quick test on a repository secret.
> 
> This [PR](https://github.com/jhesketh/forktest/pull/4) is open from a branch within the repository and is able to use the secret as seen by the *** in the output: https://github.com/jhesketh/forktest/actions/runs/6258522449/job/16992691391?pr=4#step:2:5
> 
> This [PR](https://github.com/jhesketh/forktest/pull/5) is opened from a fork, but does not have access to the secret (it is blank): https://github.com/jhesketh/forktest/actions/runs/6258527574/job/16992706304?pr=5#step:2:5
> 
> Due to the sensitivity of this, we should test the desired behavour in this repository.

I've tested this once again for this repository. My findings are consistent with yours:
- external contributor has no access to any secrets, other than GITHUB_TOKEN
- GITHUB_TOKEN token is autogenerated and is not stored anywhere. It is a token that is only valid for the duration of the workflow it was created for.
- GITHUB_TOKEN has "read" permissions when running on PRs from external forks: https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

**Comment by pstibrany** [comment]
> However, if someone who isn't a maintainer or contributor opens a PR to modify the build image, how do we get that image built? Would we need a maintainer to manually trigger the build, and would we still have the same restrictions given the code would still be coming from a fork?

We need to build it manually, or send our own PR. In practice we don't get many external contributions to build-image, so this is acceptable tradeoff.

**Pull request merged**
